Добавление получение рандомных объектов

ProtectorRTD opened the pull request:
# Добавление получение рандомных объектов

## Randomizable
Был добавлен интерфейс `Randomizable`, добавен чтобы таким образом разделять Filter-ы у которых предусматривается логика.  Также в будущем, это позволит расширять логику.

## RandomUtils
Был создан класс, который отвечает за работу с генерации случайных чисел. Он обладает лишь одним пока методом `getRandomObjectListFromList`, в будещем я планирую расширить этот класс, так чтобы избежать дублирование кода в контроллерах. Суть метода заключается в получение списка объектов, которые подходят под фильтр и уже потом из этого списка доставание случайных объектов. Сделан заранее возможность случайного количества (то есть получение списка из 2 рандомных объектов). 

Дальше уже добавление логики в Controller

ProtectorRTD committed:
1) Добавление библиотеки для маппинга

2) Начало работы с маппером

ProtectorRTD committed:
Добавление методов в маппере, а также добавление маппера в сервис

ProtectorRTD committed:
- Добавление библиотеки, для генерации impl Mapper

- А также добавление NameApi mapper

ProtectorRTD committed:
- Обновление библиотеки до последней версии

- Обработка скоростей

ProtectorRTD committed:
- Маппинг языков

ProtectorRTD committed:
- Добавление метода, которые делает маппинг фитов

ProtectorRTD committed:
Расширение ДТО объекта, добавление ActionDetailRequest

ProtectorRTD committed:
Разделение одного маппера, на много мелких мапперов

1) Для удобства тестирование
2) Для упрощения редактирование в случае багов
3) Для удобства поддержки

ProtectorRTD committed:
- Добавлен маппер для легендарных действий с зацепом под будущее обновление

 - исправлены орфографические ошибки

ProtectorRTD committed:
- Обновление DTO объекта, теперь все действия внесены в категорию ActionDetailRequest

ProtectorRTD committed:
- Обновление DTO для использования корректных тегов

- Добавление мапперов Tag-ов

ProtectorRTD committed:
- Добавление аннотации Билдер для сущности Skill

- Создание маппера для Skill

ProtectorRTD committed:
- Добавление LairMapper

ProtectorRTD committed:
- добавление маппинга источника

- добавление маппинга environment-а

ProtectorRTD committed:
- Добавление маппинга NPC

- Добавление маппинга Чувств

ProtectorRTD committed:
- Добавление ошибки, которая используется при случаях, когда маппинг был сделан неправильно

- Использование ошибки в Switch statement-ах

ProtectorRTD committed:
-Исправление пути

ProtectorRTD committed:
Исправление в структурах мапперов, корректировка главного маппера Бестиария

ProtectorRTD committed:
- подправка табуляции
- изменение название класса

ProtectorRTD committed:
Добавление класса RandomUtill, суть которого является генерация и наполнение списком рандомных объектов из того что было взято из бд на основе фильров

Добавлен интерфейс, для упрощенной работы, благодаря ему можно маркировать те Filter классы, чья рандомизация поддерживается

ProtectorRTD committed:
Добавление интерфейса к ItemFilter-u

Добавление в ApiController возможности рандом генерации

Ренейм класса утилитицы

ProtectorRTD committed:
Добавление рандома в ItemApiController

ProtectorRTD committed:
Revert "Исправление в структурах мапперов, корректировка главного маппера Бестиария"

This reverts commit bcb40ea08d2be44cb6684a251db0e49bd0316103.

ProtectorRTD committed:
Revert "-Исправление пути"

This reverts commit 46dab75cfe504f56dd5a24f0d5982cd3db71e137.

ProtectorRTD committed:
Revert "- Добавление ошибки, которая используется при случаях, когда маппинг был сделан неправильно"

This reverts commit bf77343528f75c75dfd5c6088e9ea27b54c086a7.

ProtectorRTD committed:
Revert "- Добавление маппинга NPC"

This reverts commit 8f2c07af4958f1161882a32efb2c64f68f9c19cb.

ProtectorRTD committed:
Revert "- добавление маппинга источника"

This reverts commit c56b730927020209231c2806ab261f873cb4ce5e.

ProtectorRTD committed:
Revert "- Добавление LairMapper"

This reverts commit 09b0e536af0b0a83379390d389ad70ff846e96cd.

ProtectorRTD committed:
Revert "- Добавление аннотации Билдер для сущности Skill"

This reverts commit ca5f2b88065ea488d9f5688c83b2ccd2d5a384c4.

ProtectorRTD committed:
Revert "- Обновление DTO для использования корректных тегов"

This reverts commit d735b561010459907563632ff73363ae4ebbad09.

ProtectorRTD committed:
Revert "- Обновление DTO объекта, теперь все действия внесены в категорию ActionDetailRequest"

This reverts commit 524606d5b6159ff7bcc5a9df83bfe5c1c6650c91.

ProtectorRTD committed:
Revert " - Добавлен маппер для легендарных действий с зацепом под будущее обновление"

This reverts commit e788d1fb32e281a50b52bcfb995a26434bc99423.

ProtectorRTD committed:
Revert "Разделение одного маппера, на много мелких мапперов"

This reverts commit 87e3e78f82e733b8e8987b9237ddc5d512be20c3.

ProtectorRTD committed:
Revert "Расширение ДТО объекта, добавление ActionDetailRequest"

This reverts commit 484e0702a49aaea16be206ca5d346ed0fe923738.

ProtectorRTD committed:
Revert "- Добавление метода, которые делает маппинг фитов"

This reverts commit 4d2349487b545704ce38f1eef930cf48ab80748b.

ProtectorRTD committed:
Revert "- Маппинг языков"

This reverts commit e7bf4e8298027732366e3ca69e83657778cfcb53.

ProtectorRTD committed:
Revert "- Обновление библиотеки до последней версии"

This reverts commit 48083559548ae3ea1eb491502b5149bf31c906e1.

ProtectorRTD committed:
Revert "- Добавление библиотеки, для генерации impl Mapper"

This reverts commit c8fa16ee86b604110e6addb4abab8dc7f92b1b23.

ProtectorRTD committed:
Revert "Добавление методов в маппере, а также добавление маппера в сервис"

This reverts commit ad782dbaced835f71a9be7a97438fc607ff2c8b2.

ProtectorRTD committed:
Revert "1) Добавление библиотеки для маппинга"

This reverts commit 82fae1282709a499c7fc16c88bf93ab957301bce.

ProtectorRTD committed:
Merge branch 'dev' into feature/item-magic-random

Magistrus commented on src/main/java/club/dnd5/portal/controller/api/item/ItemApiController.java:
```diff
@@ -76,6 +80,15 @@ public List<ItemApi> getItem(@RequestBody ItemRequestApi request) {
+			if (request.getFilter().getRandom()) {
+				Pageable pageable = PageAndSortUtil.getPageable(request);
+				int sizeList = pageable.getPageSize();
+				return RandomUtils.getRandomObjectListFromList(itemRepository.findAll(specification)
```
не нравится мне несколько ретурнов в одном методе, усложняет логику. Но как по другому сделать не знаю

ProtectorRTD replied:
Я подумаю ещё над этим

Magistrus commented on src/main/java/club/dnd5/portal/util/RandomUtils.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.*;
+
+public class RandomUtils {
+	private static Random random = new Random();
```
final

ProtectorRTD committed:
Добавление final

ProtectorRTD committed:
Merge remote-tracking branch 'origin/feature/item-magic-random' into feature/item-magic-random

ProtectorRTD committed:
Merge branch 'dev' into feature/item-magic-random

svifty7 commented on src/main/java/club/dnd5/portal/util/RandomUtils.java:
```diff
@@ -0,0 +1,30 @@
+		while (resultList.size() < sizeList) {
+			int randomIndex = random.nextInt(entityList.size());
+			// Проверка находится ли уже под таким индексом объект (то есть не рандомили мы объект, который уже есть)
+			if (!selectedIndices.contains(randomIndex)) {
+				resultList.add(entityList.get(randomIndex));
+				selectedIndices.add(randomIndex);
+			}
```
не знаю как принято в java, но, возможно, так будет симпатичнее:
```java
while (resultList.size() < sizeList) {
  int randomIndex = random.nextInt(entityList.size());
  if (selectedIndices.contains(randomIndex)) {
    continue;
  }
  resultList.add(entityList.get(randomIndex));
  selectedIndices.add(randomIndex);
}
```

svifty7 commented on src/main/java/club/dnd5/portal/util/RandomUtils.java:
```diff
@@ -0,0 +1,30 @@
+			throw new ApiException(HttpStatus.BAD_REQUEST, "The list is empty");
+		}
+		// Чтобы избегать дубликатов в коллекции
+		Set<Integer> selectedIndices = new HashSet<>();
```
а почему тут `Integer`, а `randomIndex` — `int`?

ProtectorRTD replied:
потому что wrapper class-ы есть в jave

svifty7 commented on src/main/java/club/dnd5/portal/util/RandomUtils.java:
```diff
@@ -0,0 +1,30 @@
+
+	private RandomUtils() {}
+
+	public static <T> List<T> getRandomObjectListFromList(List<T> entityList, int sizeList) {
```
`sizeList` предлагаю заменить на `count` или `listSize`, т.к. текущий вариант переводится как "список размеров"

ProtectorRTD replied:
согласен исправлю

ProtectorRTD commented on src/main/java/club/dnd5/portal/util/RandomUtils.java:
```diff
@@ -0,0 +1,30 @@
+			throw new ApiException(HttpStatus.BAD_REQUEST, "The list is empty");
+		}
+		// Чтобы избегать дубликатов в коллекции
+		Set<Integer> selectedIndices = new HashSet<>();
```
нет, коллекция не так работает

ProtectorRTD committed:
Обновление логики контроллера

Изменение название аргумента

ProtectorRTD committed:
**Добавление** абстракции

Были добавлен интерфейс SpecificationCommon
для того чтобы можно было создавать Спецификацию, в будущем сервис можно дополнять, чтобы больше выносить кода именно туда, а также в целом можно расширить дерево, какие Спецификации нужны и какие методы для этого

Конкретно под эту задачу, я расширил Randomizable, и добавил в него методы связанные с категориями и книгами, чтобы потом их вынести в сервис

Весь дублирующий код спецификации был удален и заменем на использование из сервиса

ProtectorRTD committed:
Merge branch 'dev' into feature/item-magic-random

Swatgy commented on src/main/java/club/dnd5/portal/util/RandomUtils.java:
```diff
@@ -0,0 +1,30 @@
+
+	private RandomUtils() {}
+
+	public static <T> List<T> getRandomObjectListFromList(List<T> entityList, int listSize) {
```
А обязательно ли нам здесь иметь `List`? Если необходимости объективной нет, я бы заменил на тот же `Set` или `Collection`. Тогда код внутри можно было бы упростить и просто добавлять покуда не заполнится сет.

Swatgy commented on src/main/java/club/dnd5/portal/util/RandomUtils.java:
```diff
@@ -0,0 +1,30 @@
+
+	private RandomUtils() {}
+
+	public static <T> List<T> getRandomObjectListFromList(List<T> entityList, int listSize) {
```
В текущей имплементации есть проблема того, что если желаемый размер больше размера сорсов мы уходим в бесконечный луп. Я бы рекомендовал добавить проверку что он хотя бы такого же размера.

ProtectorRTD replied:
Хорошее замечание, подправлю в ближайшие дни

Magistrus committed:
Merge branch 'dev' into feature/item-magic-random

ProtectorRTD committed:
Merge branch 'dev' into feature/item-magic-random
